mm2src/mm2_bitcoin/keys/src/address.rs

```diff
@@ -281,48 +306,68 @@ impl Address {
                 self.prefix, network_addr_prefixes.p2pkh, network_addr_prefixes.p2sh
             ));
         };
-        CashAddress::new(network_prefix, self.hash.to_vec(), address_type)
+        CashAddress::new(network_prefix, self.locking_destination.to_vec(), address_type)
     }
 
     pub fn from_segwitaddress(segaddr: &str, checksum_type: ChecksumType) -> Result<Address, String> {
         let address = SegwitAddress::from_str(segaddr).map_err(|e| e.to_string())?;
 
-        let (script_type, mut hash) = if address.program.len() == 20 {
-            (AddressScriptType::P2WPKH, AddressHashEnum::default_address_hash())
-        } else if address.program.len() == 32 {
-            (AddressScriptType::P2WSH, AddressHashEnum::default_witness_script_hash())
-        } else {
-            return Err("Expect either 20 or 32 bytes long hash".into());
+        let (script_type, mut locking_destination) = match (address.version_as_u8(), address.program.len()) {
+            (0, 20) => (AddressScriptType::P2WPKH, LockingDestination::default_address_hash()),
+            (0, 32) => (
+                AddressScriptType::P2WSH,
+                LockingDestination::default_witness_script_hash(),
+            ),
+            (0, _) => return Err("Expect either 20 or 32 bytes long hash for witness v0 address".into()),
+            (1, 32) => (
+                AddressScriptType::P2TR,
+                LockingDestination::default_tweaked_xonly_pubkey(),
+            ),
+            (1, _) => return Err("Expect 32 bytes long public key for witness v1 address".into()),
+            (v, _) => return Err(format!("Unsupported segwit version: {v}")),
+        };
+
+        locking_destination.copy_from_slice(address.program.as_slice());
+
+        let addr_format = AddressFormat::Segwit {
+            version: address.version_as_u8(),
         };
-        hash.copy_from_slice(address.program.as_slice());
 
         let hrp = Some(address.hrp);
 
         Ok(Address {
             prefix: AddressPrefix::default(),
-            hash,
+            locking_destination,
             checksum_type,
             hrp,
             pubkey: None,
-            addr_format: AddressFormat::Segwit,
+            addr_format,
             script_type,
         })
     }
 
     pub fn to_segwitaddress(&self) -> Result<SegwitAddress, String> {
-        match &self.hrp {
-            Some(hrp) => Ok(SegwitAddress::new(&self.hash, hrp.to_string())),
-            None => Err("hrp must be provided for segwit address".into()),
+        match (&self.hrp, &self.addr_format) {
+            (Some(hrp), AddressFormat::Segwit { version }) => {
+                Ok(SegwitAddress::new(&self.locking_destination, hrp.to_string(), *version)
+                    .map_err(|e| e.to_string())?)
+            },
+            _ => Err("hrp and segwit version must be provided for segwit address".into()),
```

Comment:
It's better to break this logic into pieces, like:

(dummy code):

```rs
let Some(hrp) = self.hrp else {
   return Err("hrp must be provided for segwit address");
};

if let AddresssFormat::Segwit ... = self.addr_format {
   ...
} else {
   Err("Invalid address format, expected Segwit, found {}", self.addr_format)
}
```

This will be much easier to debug in case of `to_segwitaddress` being failed.